### ISSUE TITLE
Custom fonts in a jar

### ISSUE THREAD

**Issue by smogg** [clj-pdf/clj-pdf#152]
I'm trying to load custom fonts while running my project as packaged jar:

```
(g2d/g2d-register-fonts [[(-> "public/fonts/" io/resource .getPath) false]])
```

That doesn't seem to be working (probably cause fonts are bytestreams now?).
Does `clj-pdf` currently supports this (without creating temp dirs to give it an actual _file_ or something like that?)

**Comment by yogthos** [comment]
The fonts are looked up relative to the filesystem by [this](https://github.com/clj-pdf/clj-pdf/blob/14b6fb87b33ee5f2917c0b7a48c8a79e80386044/src/java/cljpdf/text/FontFactoryImp.java#L164) piece of code that's called from [here](https://github.com/clj-pdf/clj-pdf/blob/94358dcd67ed1da7e414e859a7539c3954ec675a/src/clj/clj_pdf/core.clj#L96) when the font is resolved. So, it doesn't look like there's a way to read the font directly from the jar unfortunately.

**Comment by smogg** [comment]
Hmm... I tracked it down to this if statement breaking the chain:

https://github.com/clj-pdf/clj-pdf/blob/14b6fb87b33ee5f2917c0b7a48c8a79e80386044/src/java/cljpdf/text/pdf/DefaultFontMapper.java#L273

It gets an url in an absolute format: `file:/Users/username/project-name/target/uberjar/project-name-0.1.0-SNAPSHOT-standalone.jar!/public/fonts/`

Is there something we could maybe change in this code or am I totally off? I'm not familiar with Java.

**Comment by yogthos** [comment]
Yeah it should be possible to add support for reading from an input stream, however that part of the code only deals with reading the filenames. Also need to track down places where the content gets read when the font is used. I'll try take a look when I get a chance, but if you're feeling brave in the meantime I can help with a PR as well.

**Comment by smogg** [comment]
Sure, if you've got patience for it then I'm happy to give it a try. I'll dig deeper tomorrow. Thanks

**Comment by yogthos** [comment]
sounds like a plan

**Comment by smogg** [comment]
I think I may have misdiagnosed the problem from the beginning.
Here's what I was trying to do originally:

```clojure
(g2d/g2d-register-fonts [[(-> "fonts/" io/resource .getPath) false]])

;; later when generating the pdf
[:graphics {:under true :translate [0 200]}
       (fn [g2d]
         (doto g2d
           (.setFont (java.awt.Font. "Strato" java.awt.Font/PLAIN 20))
           (.drawString "Testing Graphics" 20 160)))]
```

And it worked with `lein run` but not with `java -jar ...`. 
What I missed is the fact that I had the "Strato" font installed in the system, which is why the `java.awt.Font.` was able to pick it up.
I presume this is not the right way to use custom fonts with `[:graphics]` element?

As for other elements, ie: `[:paragraph {:ttf-name "...."}]`, it appears to be working even from a jar.
I created an example project here https://github.com/smogg/custom-font-test
The `output/` folder contains a PDF created with `lein run` and with `java -jar ...` (from outside of the project dir to make sure `resources/fonts` folder is not available).

When you run the test project I also print `(g2d/get-font-maps)`, which correctly displays fonts when `lein run` (even tho my `[:graphics]` still looks like a default font) and prints empty map I run the jar.

**Comment by yogthos** [comment]
Ah so that's mostly good news then if it's a problem specific to the graphics element. The `java.awt.Font` construction looks right, but perhaps the fonts for awt have to be registered differently. You could try using:
```
(java.awt.Font/createFont java.awt.Font/TRUETYPE_FONT font-input-stream)
```

**Comment by smogg** [comment]
Yeah, I've came up with similar code:

```
    (defn graphics-env []
      (.. java.awt.GraphicsEnvironment
          getLocalGraphicsEnvironment))

    (let [font (java.awt.Font/createFont java.awt.Font/TRUETYPE_FONT (io/input-stream (io/resource "fonts/Strato-linked.ttf")))]
      (.registerFont (graphics-env) font)
      (println (.getAvailableFontFamilyNames (graphics-env )))) ;; font is available
```

But using it by name, later on, doesn't seem to be working. But I think we are getting closer. It's late in my timezone so I'll continue investigating tomorrow.

**Comment by smogg** [comment]
Ok, I feel like this is above me. I've tried a lot of things, I see fonts registered but have no idea how to use them without calling the `DefaultFontMapper/insertDirectory` method, which doesn't work when running from a jar. I tried creating new java.awt.Font's like above, but they don't work if not registered. I tried using `FontFactory/getFont` but `.setFont` expects `java.awt.font` type. I give up for now and will probably use a temp directory with `g2d-register-fonts` unless you have some pointers?

**Comment by yogthos** [comment]
I can't really think of much else that you could do here unfortunately.
